io/v7/rfile/test/rfile.cxx

```diff
@@ -0,0 +1,568 @@
+   hist->FillRandom("gaus", 1000);
+
+   auto file = RFile::Recreate(fileGuard.GetPath());
+   file->Put("hist", *hist);
+   EXPECT_TRUE(file->Get<TH1D>("hist"));
+
+   EXPECT_EQ(ROOT::GetROOT()->GetListOfFiles()->GetSize(), 0);
+}
+
+TEST(RFile, WriteInvalidPaths)
+{
+   FileRaii fileGuard("test_rfile_write_invalid.root");
+
+   auto file = RFile::Recreate(fileGuard.GetPath());
+   std::string a;
+   EXPECT_THROW(file->Put("", a), ROOT::RException);
+   EXPECT_THROW(file->Put("..", a), ROOT::RException);
+   EXPECT_THROW(file->Put(" a", a), ROOT::RException);
+   EXPECT_THROW(file->Put("a\n", a), ROOT::RException);
+   EXPECT_THROW(file->Put(".", a), ROOT::RException);
+   EXPECT_THROW(file->Put("\0", a), ROOT::RException);
+   EXPECT_THROW(file->Put(".a", a), ROOT::RException);
+   EXPECT_THROW(file->Put("a..", a), ROOT::RException);
+}
+
+TEST(RFile, OpenForUpdating)
+{
+   FileRaii fileGuard("test_rfile_update.root");
+
+   {
+      TH1D hist("hist", "", 100, -10, 10);
+      hist.FillRandom("gaus", 1000);
+      auto file = RFile::Recreate(fileGuard.GetPath());
+      file->Put("hist", hist);
+   }
+
+   auto file = RFile::OpenForUpdate(fileGuard.GetPath());
+   EXPECT_TRUE(file->Get<TH1D>("hist"));
+   {
+      auto hist2 = std::make_unique<TH1D>("hist2", "a different hist", 10, -1, 1);
+      file->Put("hist2", *hist2);
+   }
+   EXPECT_TRUE(file->Get<TH1D>("hist2"));
+
+   EXPECT_EQ(ROOT::GetROOT()->GetListOfFiles()->GetSize(), 0);
+}
+
+TEST(RFile, PutOverwrite)
+{
+   FileRaii fileGuard("test_rfile_putoverwrite.root");
+
+   auto file = RFile::Recreate(fileGuard.GetPath());
+
+   {
+      TH1D hist("hist", "", 100, -10, 10);
+      hist.FillRandom("gaus", 1000);
+      file->Put("hist", hist);
+   }
+
+   {
+      auto hist = file->Get<TH1D>("hist");
+      ASSERT_TRUE(hist);
+      EXPECT_EQ(static_cast<int>(hist->GetEntries()), 1000);
+   }
+
+   // Try putting another object at the same path, should fail
+   TH1D hist2("hist2", "a different hist", 10, -1, 1);
+   hist2.FillRandom("gaus", 10);
+   EXPECT_THROW(file->Put("hist", hist2), ROOT::RException);
+
+   // Try with Overwrite, should work (and preserve the old object)
+   file->Overwrite("hist", hist2);
+   {
+      auto hist = file->Get<TH1D>("hist");
+      ASSERT_TRUE(hist);
+      EXPECT_EQ(static_cast<int>(hist->GetEntries()), 10);
+
+      hist = file->Get<TH1D>("hist;1");
+      ASSERT_TRUE(hist);
+      EXPECT_EQ(static_cast<int>(hist->GetEntries()), 1000);
+   }
+
+   // Now try overwriting without preserving the existing object
+   std::string s;
+   file->Overwrite("hist", s, false);
+   {
+      // the previous cycle should be gone...
+      auto hist = file->Get<TH1D>("hist;2");
+      EXPECT_EQ(hist, nullptr);
+      // ...but any cycle before the latest should still be there!
+      hist = file->Get<TH1D>("hist;1");
+      EXPECT_NE(hist, nullptr);
+   }
+}
+
+TEST(RFile, WrongExtension)
+{
+   FileRaii fileGuard("test_rfile_wrong.xml");
```

Comment:
Until now we have not enforced the convention for the extension.

Reply:
But if I'm not mistaken TFile will try to open a TXMLFile if it detects this extension (at least with TFile::Open) and I wanted to avoid that

Reply:
That is a slightly different interface per say.  `TFile::Open` (as oppose to the TFile constructor) is supposed to find the right plugin and it indeed bases that search on pattern matching (`protocol` ... filename `.extension`) but if some code is creating the pluging class directly then there is less checks (in part because they are redundant, in part because `new TXMLFile(filename)` is an explicit request to open that file as xml)